go-controller/pkg/ovn/controller/services/services_controller.go

```diff
@@ -0,0 +1,484 @@
+		klog.V(2).InfoS("Error syncing service, retrying", "service", klog.KRef(ns, name), "err", err)
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	klog.Warningf("Dropping service %q out of the queue: %v", key, err)
+	c.queue.Forget(key)
+	utilruntime.HandleError(err)
+}
+
+func (c *Controller) syncServices(key string) error {
+	startTime := time.Now()
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	klog.Infof("Processing sync for service %s on namespace %s ", name, namespace)
+	metrics.MetricSyncServiceCount.WithLabelValues(key).Inc()
+
+	defer func() {
+		klog.V(4).Infof("Finished syncing service %s on namespace %s : %v", name, namespace, time.Since(startTime))
+		metrics.MetricSyncServiceLatency.WithLabelValues(key).Observe(time.Since(startTime).Seconds())
+	}()
+
+	service, err := c.serviceLister.Services(namespace).Get(name)
+	// It´s unlikely that we have an error different that "Not Found Object"
+	// because we are getting the object from the informer´s cache
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	// Delete the Service from OVN if:
+	// - it was deleted from the cache (doesn't exist in Kubernetes anymore)
+	// - it mutated to a new service Type that we don't handle (ExternalName, Headless)
+	if err != nil || !util.ServiceTypeHasClusterIP(service) || !util.IsClusterIPSet(service) {
+		// Obtain the VIPs associated to the Service from the Service Tracker
+		vips := c.serviceTracker.getService(name, namespace)
+		for vipKey := range vips {
+			// the VIP is stored with the format IP:Port/Protocol
+			vip, proto := splitVirtualIPKey(vipKey)
+			// Get the OVN LoadBalancer
+			lbID, err := loadbalancer.GetOVNKubeLoadBalancer(proto)
+			if err != nil {
+				klog.Errorf("Error getting OVN LoadBalancer for protocol %s", proto)
+				return err
+			}
+			// Delete the Service VIP from OVN
+			klog.Infof("Deleting service %s on namespace %s from OVN", name, namespace)
+			if err := loadbalancer.DeleteLoadBalancerVIP(lbID, vip); err != nil {
+				klog.Errorf("Error deleting VIP %s on OVN LoadBalancer %s", vip, lbID)
+				return err
+			}
+			// Delete the Service VIP from the Service Tracker
+			c.serviceTracker.deleteServiceVIP(name, namespace, vip, proto)
+		}
+		// Delete the Service form the Service Tracker
+		c.serviceTracker.deleteService(name, namespace)
+		return nil
+	}
+	klog.Infof("Creating service %s on namespace %s on OVN", name, namespace)
+	// The Service exists in the cache: update it in OVN
+	// Get the endpoint slices associated to the Service
+	esLabelSelector := labels.Set(map[string]string{
+		discovery.LabelServiceName: name,
+	}).AsSelectorPreValidated()
+	endpointSlices, err := c.endpointSliceLister.EndpointSlices(namespace).List(esLabelSelector)
+	if err != nil {
+		// Since we're getting stuff from a local cache, it is basically impossible to get this error.
+		c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToListEndpointSlices",
```

Comment:
Nit question: should we be triggering events for these errors? They would qualify as internal errors to ovn-kube, and if I understand events right: it's a construct to notify the user of something, so that he can act upon it?

Reply:
the user need to know that his services doesn´t work because of X ... in this case he is going to check his services and endpoints with kubectl and everything will be correct.
We should not force them to go directly to ovnkube, and this way kubernetes can warn him of the problem